src/target/riscv/riscv.c

```diff
@@ -604,90 +606,349 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
+	return tdata1;
+}
+
+static riscv_reg_t set_match_ge_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_ge_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_GE);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_eq_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
 	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
-	if (trigger->length == 1) {
-		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_8BIT & 3);
-		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_8BIT >> 2) & 3);
-	}
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	tdata2 = trigger->address;
-	ret = set_trigger(target, idx, tdata1, tdata2, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-	if (ret != ERROR_OK)
-		return ret;
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	return ERROR_OK;
+	return tdata1;
+}
+
+static riscv_reg_t set_match_eq_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_EQUAL);
+	return tdata1;
+}
+
+static struct match_triggers_field_info fill_match_triggers_field_info_t2(
+		struct target *target, struct trigger *trigger)
+{
+	struct match_triggers_field_info result = {
+		.common = set_common_t2(target, trigger),
+		.size = {
+			.any = set_size_any_t2(),
+			.s8bit = set_size_s8bit_t2()
+		},
+		.chain = {
+			.enable = set_chain_enable_t2(),
+			.disable = set_chain_disable_t2()
+		},
+		.match = {
+			.napot = set_match_napot_t2(),
+			.lt = set_match_lt_t2(),
+			.ge = set_match_ge_t2(),
+			.eq = set_match_eq_t2()
+		},
+		.tdata1_ignore_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target))
+	};
+	return result;
+}
+
+static struct match_triggers_field_info fill_match_triggers_field_info_t6(
+		struct target *target, struct trigger *trigger)
+{
+	struct match_triggers_field_info result = {
+		.common = set_common_t6(target, trigger),
+		.size = {
+			.any = set_size_any_t6(),
+			.s8bit = set_size_s8bit_t6()
+		},
+		.chain = {
+			.enable = set_chain_enable_t6(),
+			.disable = set_chain_disable_t6()
+		},
+		.match = {
+			.napot = set_match_napot_t6(),
+			.lt = set_match_lt_t6(),
+			.ge = set_match_ge_t6(),
+			.eq = set_match_eq_t6()
+		},
+		.tdata1_ignore_mask = 0
+	};
+	return result;
+}
+
+static int maybe_add_trigger_t2_t6(struct target *target,
+		struct trigger *trigger, struct match_triggers_field_info set)
+{
+	int ret = ERROR_OK;
+
+	if (!trigger->execute && trigger->length > 1) {
+		if (can_use_napot_match(trigger)) {
+			LOG_TARGET_DEBUG(target, "trying t request NAPOT match");
+			struct trigger_request_info napot = {
+				.tdata1 = set.common | set.size.any | set.chain.disable |
+					set.match.napot,
+				.tdata2 = trigger->address | ((trigger->length - 1) >> 1),
+				.tdata1_ignore_mask = set.tdata1_ignore_mask
+			};
```

Comment:
When `trigger_type` is passed to `maybe_add_trigger_t2_t6()`, as suggested in another comment, the creation of structures like `napot`, `ge_1`, `lt_1` ... can be simplified this way:

```suggestion
			struct trigger_request_info napot = fill_trigger_request_info_napot(
					trigger_type, false, trigger->address | ((trigger->length - 1) >> 1));
```

Function `fill_trigger_request_info_napot()` (and `*_gt()`, `*_lt()`, `*_eq()`) can look like this (pseudocode):

```
struct trigger_request_info fill_trigger_request_info_napot(unsigned int trigger_type, struct trigger *trigger, bool is_chained, riscv_reg_t tdata2)
{
	riscv_reg_t tdata1;
	riscv_reg_t tdata1_ignore_mask;	
	if (trigger_type == 2) {
		tdata1 = fill_common_tdata1_t2(trigger);
		/* ... */
	} else {
		tdata1 = fill_common_tdata1_t6(trigger);
		/* ... */	
	}
	struct trigger_request_info result = {
		.tdata1 = tdata1, .tdata2 = tdata2, .tdata1_ignore_mask = tdata1_ignore_mask
	};
	return result;
}
```

The relatively complex (nested) `struct match_triggers_field_info` will then not be needed.

Reply:
Thank you for the suggestion. I would like to keep the current approach. Now there is no `if`s dependent on trigger type since it is known on the call site `maybe_add_trigger_t2_t6`. If I were to follow your suggestion, it would create many `if`s with the same condition, so, in my opinion, it would complicate control flow and the code would become more error-prone.

Now:
```
static int maybe_add_trigger_t2_t6(struct target *target,
        struct trigger *trigger, struct match_triggers_field_info set)
{
   /*...*/
    if (!trigger->execute && trigger->length > 1) {
         /* Prepare corresponding structure */
         ret = try_setup_single_match_trigger(target, trigger, napot);
        if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
            return ret;
        /* Prepare corresponding structures */
        ret = try_setup_chained_match_triggers(target, trigger, ge_1, lt_2);
        if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
            return ret;
        /* Prepare corresponding structures */
        ret = try_setup_chained_match_triggers(target, trigger, lt_1, ge_2);
        if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
            return ret;
    }
    /* Prepare corresponding structure */
    return try_setup_single_match_trigger(target, trigger, eq);
}
```

Your suggestion:
```
struct trigger_request_info fill_trigger_request_info_napot(unsigned int trigger_type, ...)
{
	if (trigger_type == 2) {
		/* ... */
	} else {
		/* ... */	
	}
       /*...*/
}

struct trigger_request_info fill_trigger_request_info_ge(unsigned int trigger_type, ...)
{
	if (trigger_type == 2) {
		/* ... */
	} else {
		/* ... */	
	}
       /*...*/
}

struct trigger_request_info fill_trigger_request_info_lt(unsigned int trigger_type, ...)
{
	if (trigger_type == 2) {
		/* ... */
	} else {
		/* ... */	
	}
       /*...*/
}

struct trigger_request_info fill_trigger_request_info_eq(unsigned int trigger_type, ...)
{
	if (trigger_type == 2) {
		/* ... */
	} else {
		/* ... */	
	}
       /*...*/
}

static int maybe_add_trigger_t2_t6(struct target *target, struct trigger *trigger, unsigned int trigger_type)
{
   /*...*/
    if (!trigger->execute && trigger->length > 1) {
           /* ...*/
            ret = try_setup_single_match_trigger(target, trigger, fill_trigger_request_info_napot(trigger_type, ...));
           /*...*/
        }
        /*...*/
        ret = try_setup_chained_match_triggers(target, trigger, fill_trigger_request_info_ge(trigger_type, /*chained = */ true, ...), fill_trigger_request_info_lt(trigger_type, /*chained = */ false, ...));
        /*...*/
        ret = try_setup_chained_match_triggers(target, trigger, fill_trigger_request_info_lt(trigger_type, /*chained = */ true, ...), fill_trigger_request_info_ge(trigger_type, /*chained = */ false, ...));
        /*...*/
    return try_setup_single_match_trigger(target, trigger, fill_trigger_request_info_eq(trigger_type, ...));
}
```

Reply:
Hi @en-sc, thank you for replying.

I am afraid the code comparison above is not fair. That's because the code titled `Now:` does not show the complex structure of `struct match_triggers_field_info` data type nor the code need to fill that structure.

Having the `if (trigger_type == 2) {...} else {...}` conditions inside the `fill_trigger_request_info_*` functions looks all right to me, I do not have any concern about complexity. In fact such functions will be straightforward to read, debug and would even be easily unit-testable. 

Also the last piece of code above can be made nicer by getting rid of the nested calls to `fill_trigger_request_info_*`. Instead, the results can be stored in variables:

```
        // Original:
        ret = try_setup_chained_match_triggers(target, trigger, fill_trigger_request_info_ge(trigger_type, /*chained = */ true, ...), fill_trigger_request_info_lt(trigger_type, /*chained = */ false, ...));
        
        // Suggested:
        struct trigger_request_info ge_1 = fill_trigger_request_info_ge(trigger_type, /*chained = */ true, ...);
        struct trigger_request_info lt_2 = fill_trigger_request_info_lt(trigger_type, /*chained = */ false, ...);
        ret = try_setup_chained_match_triggers(target, trigger, ge_1, lt_2);        
```

Please, if you don't have strong feeling about keeping `struct match_triggers_field_info` in the code, it would be my recommendation to simplify the code as suggested.

Reply:
I think I like where @en-sc is going with this. The complicated data structure only contains constants. It could be flattened but that wouldn't really change much.
If all these set_match_ge_t2() were inlined then I think it would be fairly compact and simple as well.